pkg/virt-handler/vm.go

```diff
@@ -2772,6 +2765,28 @@ func (d *VirtualMachineController) isHostModelMigratable(vmi *v1.VirtualMachineI
 	return nil
 }
 
+func (d *VirtualMachineController) claimKVMDeviceOwnership(vmi *v1.VirtualMachineInstance) error {
+	isolation, err := d.podIsolationDetector.Detect(vmi)
+	if err != nil {
+		return err
+	}
+
+	kvmPath := path.Join(isolation.MountRoot(), "dev", "kvm")
+	hasKVM, err := diskutils.FileExists(kvmPath)
+	if err != nil {
+		return err
+	}
+
+	// This is the logic applied by virt-launcher (see virtwrap/converter/converter.go)
```

Comment:
it seems fragile to replicate other component functionality like this. wouldn't it be nobler to define a shared utility function called from both?

Regardless, why do we need this complex functionality? Either d.clusterConfig.IsUseEmulation() is true and we don't care about /dev/kvm, or it is false and we must reown /dev/kvm. If /dev/kvm is needed but somehow missing, changing ownership would fail anyway, as it should.

```suggestion
        if !d.clusterConfig.IsUseEmulation() {
  		return diskutils.DefaultOwnershipManager.SetFileOwnership(kvmPath)
   	}
}

Reply:
Despite its misleading name, the semantics of `IsUseEmulation()` is more of a `CanUseEmulationIfNoKvm()`. For my understanding this is a deliberate design decision. So, if we retain the existing semantics, we do care if /dev/kvm is present even if `IsUseEmulation()` is true. Let's go for the noble approach.

Reply:
oh dear. you may want to rename `IsUseEmulation` -> `UseEmulationIfNoKvm` (even if we do not change the user-facing name of config, which is more complex due to backward compatibility)

Reply:
That would be noble as well 😉. Let me follow up on that in a separate PR, so that this one can be quickly merged and backported. The useEmulation misnomer is used across several different places (virt-controller, virt-handler, virt-launcher (`--use-emulation` flag)). I'd also update the user-facing config comment so at least `kubectl explain` gets it right.

Reply:
/hold

Reply:
/hold cancel